IndexError: list index out of range

Hi, I can not understand what exactly the error is. Does somebody has any idea?
```
[INFO] Running job "echo (trigger: interval[0:02:00], next run at: 2022-04-20 20:21:34 UTC)" (scheduled at 2022-04-20 20:19:34.563367+00:00)

[INFO] Started echo

[ERROR] Job "echo (trigger: interval[0:02:00], next run at: 2022-04-20 20:21:34 UTC)" raised an exception

Traceback (most recent call last):

  File "/usr/local/lib/python3.8/site-packages/apscheduler/executors/base.py", line 125, in run_job

    retval = job.func(*job.args, **job.kwargs)

  File "/app/main.py", line 121, in echo

    last_items[chat_id] = {'last_item': items[0].url, 'search_url': url}

IndexError: list index out of range
```

--- a/main.py
+++ b/main.py
@@ -58,15 +58,14 @@ def get_items_per_url(url):
     qq = requests.get(url, headers=headers)
 
     text = qq.text
-    log.info(text)
 
     articles = re.findall('<article(.*?)</article', text, re.S)
     log.info('Articles length %s' % len(articles))
     items = []
     for item in articles:
         if results := re.findall('<a.*?href="(.*?)">(.*?)</a>', item, re.S):
-            log.info("Found link")
             url, name = results[0]
+
         else:
             continue
 
@@ -89,7 +88,10 @@ def get_items_per_url(url):
         except Exception as e:
             logger.error(f'No image\n\t{item}')
             continue
-
+        log.info("image: " + image)
+        log.info("URL " + url)
+        log.info("Title " + name)
+        log.info("Price " + price)
         items.append(Item(name, price, torg, url, image))
     return items
 
@@ -123,7 +125,6 @@ def echo(update: Update, context):
 
     log.info("Get items")
     items = get_items_per_url(url)
-    log.info(items)
     for item in items:
         if chat_id in last_items and item.url == last_items[chat_id]['last_item']:
             #log.info('Breaking the loop')

--- a/main.py
+++ b/main.py
@@ -74,12 +74,6 @@ def get_items_per_url(url):
             name = soup_result[0].text
         else:
             continue
-        # log.info(soup_result)
-        # if results := re.findall('<a.*?href="(.*?)">(.*?)</a>', item, re.S):
-        #     url, name = results[0]
-
-        # else:
-        #     continue
 
         price_line = re.findall('aditem-main--middle--price">(.*?)</p>', item, re.S)
         if len(price_line) > 0:
@@ -130,7 +124,7 @@ def echo(update: Update, context):
 
     if chat_id not in last_items:
         # Nothing here, schedule
-        scheduler.add_job(echo, trigger='interval', args=(update, context), minutes=2, id=str(chat_id))
+        scheduler.add_job(echo, trigger='interval', args=(update, context), minutes=15, id=str(chat_id))
         log.info('Scheduled job')
         last_items[chat_id] = {'last_item': None, 'url': url}
 

--- a/main.py
+++ b/main.py
@@ -47,17 +47,18 @@ class Item:
 
 
 def get_items_per_url(url):
+    log = utils.get_logger()
     qq = requests.get(url)
 
     text = qq.text
-    # with open(r'C:\Users\Kleinanzeigen\Desktop\qq.html', 'r') as f:
-    #     text = f.read()
+    log.info(text)
 
     articles = re.findall('<article(.*?)</article', text, re.S)
-
+    log.info('Articles length %s' % len(articles))
     items = []
     for item in articles:
         if results := re.findall('<a.*?href="(.*?)">(.*?)</a>', item, re.S):
+            log.info("Found link")
             url, name = results[0]
         else:
             continue
@@ -79,7 +80,7 @@ def get_items_per_url(url):
             continue
 
         items.append(Item(name, price, torg, url, date, image))
-
+    print("Items size " + len(items))
     return items
 
 
@@ -110,8 +111,9 @@ def echo(update: Update, context):
         log.info('Scheduled job')
         last_items[chat_id] = {'last_item': None, 'url': url}
 
+    log.info("Get items")
     items = get_items_per_url(url)
-
+    log.info(items)
     for item in items:
         if chat_id in last_items and item.url == last_items[chat_id]['last_item']:
             #log.info('Breaking the loop')

--- a/main.py
+++ b/main.py
@@ -31,7 +31,6 @@ class Item:
         self.price = price
         self.torg = torg
         self.url = 'https://www.ebay-kleinanzeigen.de' + url
-        #self.date = date
         self.image = image
 
     def __repr__(self):
@@ -41,7 +40,6 @@ class Item:
         result = f'{self.title} - {self.price}'
         if self.torg:
             result += ' VB'
-        #result += f'\n\t{self.date}\n'
 
         result += self.url
         result += '\n'
@@ -85,9 +83,6 @@ def get_items_per_url(url):
         if prices := re.findall(r'\d+', price_line, re.S):
             price = int(prices[0])
 
-        # date = re.findall('aditem-addon">(.*?)</', item, re.S)[0].strip()
-        # if '{' in date or '<' in date:
-        #     continue
 
         try:
             image = re.findall('imgsrc="(.*?)"', item, re.S)[0].strip()

--- a/main.py
+++ b/main.py
@@ -107,7 +107,7 @@ def echo(update: Update, context):
 
     if chat_id not in last_items:
         # Nothing here, schedule
-        scheduler.add_job(echo, trigger='interval', args=(update, context), minutes=15, id=str(chat_id))
+        scheduler.add_job(echo, trigger='interval', args=(update, context), minutes=2, id=str(chat_id))
         log.info('Scheduled job')
         last_items[chat_id] = {'last_item': None, 'url': url}
 

--- a/main.py
+++ b/main.py
@@ -1,6 +1,8 @@
 import os
 import re
 import sys
+from bs4 import BeautifulSoup
+
 
 import requests
 from apscheduler.jobstores.sqlalchemy import SQLAlchemyJobStore
@@ -58,12 +60,16 @@ def get_items_per_url(url):
     qq = requests.get(url, headers=headers)
 
     text = qq.text
+    soup = BeautifulSoup(text, 'html.parser')
+
 
     articles = re.findall('<article(.*?)</article', text, re.S)
     log.info('Articles length %s' % len(articles))
     items = []
     for item in articles:
-        if results := re.findall('<a class="ellipsis" href="(.*?)">(.*?)</a>', item, re.S):
+        soup_result = soup.find_all("a", {"class": 'ellipsis'})
+        log.info(soup_result)
+        if results := re.findall('<a.*?href="(.*?)">(.*?)</a>', item, re.S):
             url, name = results[0]
 
         else:

--- a/main.py
+++ b/main.py
@@ -70,7 +70,11 @@ def get_items_per_url(url):
         else:
             continue
 
-        price_line = re.findall('<strong>(.*?)</strong>', item, re.S)[0]
+        price_line = re.findall('aditem-main--middle--price">(.*?)</p>', item, re.S)
+        if len(price_line) > 0:
+            price_line = price_line[0]
+        else:
+            price_line = 0
         torg = 'VB' in price_line
         price = None
         if prices := re.findall(r'\d+', price_line, re.S):

--- a/main.py
+++ b/main.py
@@ -64,7 +64,7 @@ def get_items_per_url(url):
 
 
     articles = re.findall('<article(.*?)</article', text, re.S)
-    log.info('Articles length %s' % len(articles))
+    log.info(f"Articles length {len(articles)}")
     items = []
     for item in articles:
         soup = BeautifulSoup(item, 'html.parser')

--- a/main.py
+++ b/main.py
@@ -91,7 +91,6 @@ def get_items_per_url(url):
             continue
 
         items.append(Item(name, price, torg, url, image))
-    print("Items size " + len(items))
     return items
 
 

--- a/main.py
+++ b/main.py
@@ -24,22 +24,22 @@ logger = utils.get_logger()
 
 
 class Item:
-    def __init__(self, title, price, torg, url, date, image):
+    def __init__(self, title, price, torg, url, image):
         self.title = title
         self.price = price
         self.torg = torg
         self.url = 'https://www.ebay-kleinanzeigen.de' + url
-        self.date = date
+        #self.date = date
         self.image = image
 
     def __repr__(self):
-        return f'{self.title} - {self.price} - {self.date}'
+        return f'{self.title} - {self.price}'
 
     def __str__(self):
         result = f'{self.title} - {self.price}'
         if self.torg:
             result += ' VB'
-        result += f'\n\t{self.date}\n'
+        #result += f'\n\t{self.date}\n'
 
         result += self.url
         result += '\n'
@@ -80,9 +80,9 @@ def get_items_per_url(url):
         if prices := re.findall(r'\d+', price_line, re.S):
             price = int(prices[0])
 
-        date = re.findall('aditem-addon">(.*?)</', item, re.S)[0].strip()
-        if '{' in date or '<' in date:
-            continue
+        # date = re.findall('aditem-addon">(.*?)</', item, re.S)[0].strip()
+        # if '{' in date or '<' in date:
+        #     continue
 
         try:
             image = re.findall('imgsrc="(.*?)"', item, re.S)[0].strip()
@@ -90,7 +90,7 @@ def get_items_per_url(url):
             logger.error(f'No image\n\t{item}')
             continue
 
-        items.append(Item(name, price, torg, url, date, image))
+        items.append(Item(name, price, torg, url, image))
     print("Items size " + len(items))
     return items
 

--- a/main.py
+++ b/main.py
@@ -48,7 +48,14 @@ class Item:
 
 def get_items_per_url(url):
     log = utils.get_logger()
-    qq = requests.get(url)
+    # Simulate browser headers
+    headers = {
+        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36',
+        'Host': 'www.ebay-kleinanzeigen.de',
+        'Accept': '*/*',
+        }
+
+    qq = requests.get(url, headers=headers)
 
     text = qq.text
     log.info(text)

--- a/main.py
+++ b/main.py
@@ -63,7 +63,7 @@ def get_items_per_url(url):
     log.info('Articles length %s' % len(articles))
     items = []
     for item in articles:
-        if results := re.findall('<a.*?href="(.*?)">(.*?)</a>', item, re.S):
+        if results := re.findall('<a class="ellipsis" href="(.*?)">(.*?)</a>', item, re.S):
             url, name = results[0]
 
         else:
@@ -91,7 +91,6 @@ def get_items_per_url(url):
         log.info("image: " + image)
         log.info("URL " + url)
         log.info("Title " + name)
-        log.info("Price " + price)
         items.append(Item(name, price, torg, url, image))
     return items
 

--- a/main.py
+++ b/main.py
@@ -60,14 +60,15 @@ def get_items_per_url(url):
     qq = requests.get(url, headers=headers)
 
     text = qq.text
-    soup = BeautifulSoup(text, 'html.parser')
+    
 
 
     articles = re.findall('<article(.*?)</article', text, re.S)
     log.info('Articles length %s' % len(articles))
     items = []
     for item in articles:
-        soup_result = soup.find_all("a", {"class": 'ellipsis'})
+        soup = BeautifulSoup(item, 'html.parser')
+        soup_result = soup.find_all("a", {"class": 'ellipsis'})[0]
         log.info(soup_result)
         if results := re.findall('<a.*?href="(.*?)">(.*?)</a>', item, re.S):
             url, name = results[0]

--- a/main.py
+++ b/main.py
@@ -68,13 +68,18 @@ def get_items_per_url(url):
     items = []
     for item in articles:
         soup = BeautifulSoup(item, 'html.parser')
-        soup_result = soup.find_all("a", {"class": 'ellipsis'})[0]
-        log.info(soup_result)
-        if results := re.findall('<a.*?href="(.*?)">(.*?)</a>', item, re.S):
-            url, name = results[0]
-
+        soup_result = soup.find_all("a", {"class": 'ellipsis'})
+        if len(soup_result) > 0:
+            url = soup_result[0]['href']
+            name = soup_result[0].text
         else:
             continue
+        # log.info(soup_result)
+        # if results := re.findall('<a.*?href="(.*?)">(.*?)</a>', item, re.S):
+        #     url, name = results[0]
+
+        # else:
+        #     continue
 
         price_line = re.findall('aditem-main--middle--price">(.*?)</p>', item, re.S)
         if len(price_line) > 0:

--- a/main.py
+++ b/main.py
@@ -79,7 +79,7 @@ def get_items_per_url(url):
         if len(price_line) > 0:
             price_line = price_line[0]
         else:
-            price_line = 0
+            price_line = "0"
         torg = 'VB' in price_line
         price = None
         if prices := re.findall(r'\d+', price_line, re.S):

--- a/requirements.txt
+++ b/requirements.txt
@@ -3,6 +3,6 @@ requests
 apscheduler
 SQLAlchemy
 CMRESHandler
-
+bs4
 # Dependencies for Telegram bot
 python-telegram-bot==12.3.0

